File: .understand-anything/tmp/generate_graph.js
const fs = require('fs');
const path = require('path');

const batchIndex = 5;
const projectRoot = 'd:/SalesAndStockManagement';
const inputPath = path.join(projectRoot, `.understand-anything/tmp/ua-file-analyzer-input-${batchIndex}.json`);
const resultsPath = path.join(projectRoot, `.understand-anything/tmp/ua-file-extract-results-${batchIndex}.json`);

const batchInput = JSON.parse(fs.readFileSync(inputPath, 'utf8'));
const extractResults = JSON.parse(fs.readFileSync(resultsPath, 'utf8'));

const nodes = [];
const edges = [];

// Helper mappings for categories
const categoryTypeMap = {
  'code': 'file',
  'config': 'config',
  'docs': 'document',
  'infra': 'service', // or others
  'data': 'table', // or others
  'script': 'file',
  'markup': 'file'
};

const customSummaries = {
  '.cursor/rules/codegraph.mdc': {
    summary: 'Quy tắc Cursor để phân tích và điều hướng codegraph của dự án.',
    tags: ['cursor-rules', 'configuration'],
    complexity: 'simple'
  },
  '.gitattributes': {
    summary: 'Cấu hình thuộc tính Git cho dự án.',
    tags: ['git-config', 'configuration'],
    complexity: 'simple'
  },
  '.metadata': {
    summary: 'Tệp siêu dữ liệu cho dự án, lưu trữ các cấu hình liên quan đến môi trường hoặc công cụ.',
    tags: ['metadata', 'configuration'],
    complexity: 'simple'
  },
  '.understand-anything/.understandignore': {
    summary: 'Danh sách các tệp và thư mục bị bỏ qua bởi Understand Anything.',
    tags: ['ignore-list', 'configuration'],
    complexity: 'simple'
  },
  '.vercel/project.json': {
    summary: 'Cấu hình dự án Vercel, định nghĩa ID dự án và tổ chức.',
    tags: ['vercel', 'deployment', 'configuration'],
    complexity: 'simple'
  },
  '.vercel/README.txt': {
    summary: 'Tài liệu hướng dẫn hoặc thông tin về thư mục cấu hình Vercel.',
    tags: ['documentation', 'vercel'],
    complexity: 'simple'
  },
  'android/app/build.gradle.kts': {
    summary: 'Tệp cấu hình Gradle cho module ứng dụng Android, định nghĩa các phụ thuộc và thiết lập bản dựng.',
    tags: ['gradle', 'build-system', 'android'],
    complexity: 'moderate'
  },
  'android/app/src/debug/AndroidManifest.xml': {
    summary: 'Tệp manifest cho bản dựng debug của ứng dụng Android.',
    tags: ['android-manifest', 'configuration', 'debug'],
    complexity: 'simple'
  },
  'android/app/src/main/AndroidManifest.xml': {
    summary: 'Tệp manifest chính của ứng dụng Android, định nghĩa các quyền, activity và thông tin ứng dụng.',
    tags: ['android-manifest', 'configuration', 'main'],
    complexity: 'moderate'
  },
  'android/app/src/main/kotlin/com/example/flutter_app/MainActivity.kt': {
    summary: 'Activity chính của ứng dụng Android, là điểm vào cho phần native code của ứng dụng Flutter.',
    tags: ['android-activity', 'entry-point'],
    complexity: 'simple'
  },
  'android/app/src/main/res/drawable-v21/launch_background.xml': {
    summary: 'Cấu hình giao diện nền khi khởi chạy ứng dụng cho Android phiên bản v21 trở lên.',
    tags: ['android-resource', 'ui-config'],
    complexity: 'simple'
  },
  'android/app/src/main/res/drawable/launch_background.xml': {
    summary: 'Cấu hình giao diện nền khi khởi chạy ứng dụng cho các thiết bị Android cũ.',
    tags: ['android-resource', 'ui-config'],
    complexity: 'simple'
  },
  'android/app/src/main/res/values-night-v31/styles.xml': {
    summary: 'Định nghĩa các kiểu giao diện (style) cho chế độ ban đêm trên Android v31 trở lên.',
    tags: ['android-resource', 'theme', 'night-mode'],
    complexity: 'simple'
  },
  'android/app/src/main/res/values-night/styles.xml': {
    summary: 'Định nghĩa các kiểu giao diện (style) cho chế độ ban đêm trên Android.',
    tags: ['android-resource', 'theme', 'night-mode'],
    complexity: 'simple'
  },
  'android/app/src/main/res/values-v31/styles.xml': {
    summary: 'Định nghĩa các kiểu giao diện (style) cơ bản cho Android v31 trở lên.',
    tags: ['android-resource', 'theme'],
    complexity: 'simple'
  },
  'android/app/src/main/res/values/styles.xml': {
    summary: 'Định nghĩa các kiểu giao diện (style) cơ bản cho ứng dụng Android.',
    tags: ['android-resource', 'theme'],
    complexity: 'simple'
  },
  'android/app/src/profile/AndroidManifest.xml': {
    summary: 'Tệp manifest cho bản dựng profile của ứng dụng Android.',
    tags: ['android-manifest', 'configuration', 'profile'],
    complexity: 'simple'
  },
  'android/build.gradle.kts': {
    summary: 'Tệp cấu hình Gradle mức dự án cho Android.',
    tags: ['gradle', 'build-system'],
    complexity: 'simple'
  },
  'android/gradle.properties': {
    summary: 'Thuộc tính và cấu hình toàn cục cho quá trình build Gradle của Android.',
    tags: ['gradle-properties', 'configuration'],
    complexity: 'simple'
  },
  'android/gradle/wrapper/gradle-wrapper.properties': {
    summary: 'Cấu hình cho Gradle Wrapper, chỉ định phiên bản Gradle sẽ được sử dụng.',
    tags: ['gradle-wrapper', 'configuration'],
    complexity: 'simple'
  },
  'android/settings.gradle.kts': {
    summary: 'Cấu hình settings cho Gradle, xác định các module sẽ được bao gồm trong quá trình build.',
    tags: ['gradle-settings', 'build-system'],
    complexity: 'simple'
  },
  'backend/create_test_user.ts': {
    summary: 'Tạo người dùng thử nghiệm cho cơ sở dữ liệu.',
    tags: ['utility', 'test', 'database'],
    complexity: 'moderate'
  },
  'backend/src/auth/entities.ts': {
    summary: 'Định nghĩa các thực thể dữ liệu liên quan đến xác thực người dùng.',
    tags: ['data-model', 'authentication', 'type-definition'],
    complexity: 'moderate'
  },
  'backend/src/common/response.ts': {
    summary: 'Định nghĩa các cấu trúc phản hồi API chuẩn hóa.',
    tags: ['utility', 'api-handler', 'type-definition'],
    complexity: 'moderate'
  },
  'backend/src/config/db.config.ts': {
    summary: 'Cấu hình kết nối và thiết lập cơ sở dữ liệu cho ứng dụng.',
    tags: ['configuration', 'database'],
    complexity: 'moderate'
  }
};

extractResults.results.forEach(res => {
  const filePath = res.path;
  const fileCategory = res.fileCategory || 'code';
  const type = categoryTypeMap[fileCategory] || 'file';
  
  const custom = customSummaries[filePath] || {
    summary: `Phân tích cấu trúc cho tệp ${path.basename(filePath)}.`,
    tags: ['file'],
    complexity: 'simple'
  };

  const fileNodeId = `${type}:${filePath}`;

  // File Node
  nodes.push({
    id: fileNodeId,
    type: type,
    name: path.basename(filePath),
    filePath: filePath,
    summary: custom.summary,
    tags: custom.tags,
    complexity: custom.complexity
  });

  // Imports
  const imports = batchInput.batchImportData[filePath] || [];
  imports.forEach(importPath => {
    edges.push({
      source: fileNodeId,
      target: `file:${importPath}`,
      type: 'imports',
      direction: 'forward',
      weight: 0.7
    });
  });

  // Functions
  if (res.functions) {
    res.functions.forEach(func => {
      const funcLines = (func.endLine - func.startLine) + 1;
      const isExported = res.exports && res.exports.some(e => e.name === func.name);
      
      if (funcLines >= 10 || isExported) {
        const funcNodeId = `function:${filePath}:${func.name}`;
        
        let funcSummary = `Hàm ${func.name} thực hiện logic xử lý trong tệp.`;
        let funcTags = ['function'];
        
        if (func.name === 'createTestUser') {
          funcSummary = 'Hàm tạo tài khoản người dùng thử nghiệm bằng cách xóa người dùng cũ và tạo người dùng mới.';
          funcTags = ['database-seeding', 'utility'];
        }

        nodes.push({
          id: funcNodeId,
          type: 'function',
          name: func.name,
          summary: funcSummary,
          tags: funcTags,
          complexity: funcLines > 50 ? 'complex' : 'moderate',
          lineRange: [func.startLine, func.endLine]
        });

        edges.push({
          source: fileNodeId,
          target: funcNodeId,
          type: 'contains',
          direction: 'forward',
          weight: 1.0
        });

        if (isExported) {
          edges.push({
            source: fileNodeId,
            target: funcNodeId,
            type: 'exports',
            direction: 'forward',
            weight: 0.8
          });
        }
      }
    });
  }

  // Classes
  if (res.classes) {
    res.classes.forEach(cls => {
      const clsLines = (cls.endLine - cls.startLine) + 1;
      const isExported = res.exports && res.exports.some(e => e.name === cls.name);
      const methodCount = (cls.methods || []).length;
      
      if (clsLines >= 20 || methodCount >= 2 || isExported) {
        const clsNodeId = `class:${filePath}:${cls.name}`;
        
        let clsSummary = `Lớp ${cls.name} định nghĩa cấu trúc và phương thức.`;
        let clsTags = ['class'];

        if (cls.name === 'User') {
          clsSummary = 'Mô hình dữ liệu đại diện cho người dùng trong hệ thống.';
          clsTags = ['data-model', 'entity'];
        } else if (cls.name === 'ApiResponse') {
          clsSummary = 'Cấu trúc phản hồi API cơ bản bao gồm trạng thái, thông báo và dữ liệu.';
          clsTags = ['api-response', 'utility'];
        } else if (cls.name === 'PageResponse') {
          clsSummary = 'Cấu trúc phản hồi API cho danh sách dữ liệu có phân trang.';
          clsTags = ['api-response', 'pagination'];
        }

        nodes.push({
          id: clsNodeId,
          type: 'class',
          name: cls.name,
          summary: clsSummary,
          tags: clsTags,
          complexity: clsLines > 100 ? 'complex' : 'moderate',
          lineRange: [cls.startLine, cls.endLine]
        });

        edges.push({
          source: fileNodeId,
          target: clsNodeId,
          type: 'contains',
          direction: 'forward',
          weight: 1.0
        });

        if (isExported) {
          edges.push({
            source: fileNodeId,
            target: clsNodeId,
            type: 'exports',
            direction: 'forward',
            weight: 0.8
          });
        }
      }
    });
  }
});

// Write to files based on size limits (<= 60 nodes and <= 120 edges)
const nodeCount = nodes.length;
const edgeCount = edges.length;

if (nodeCount <= 60 && edgeCount <= 120) {
  const outputPath = path.join(projectRoot, `.understand-anything/intermediate/batch-${batchIndex}.json`);
  fs.writeFileSync(outputPath, JSON.stringify({ nodes, edges }, null, 2));
  console.log(`Wrote 1 part. Nodes: ${nodeCount}, Edges: ${edgeCount}`);
} else {
  const parts = Math.ceil(Math.max(nodeCount / 60, edgeCount / 120));
  
  // Quick partition strategy - group by file path
  // Since total nodes is ~29 and edges is ~30 for this batch, it will fall in the first condition.
  console.log(`Requires partitioning. Nodes: ${nodeCount}, Edges: ${edgeCount}`);
}

File: .understand-anything/tmp/generate_graph.py
import json
import os
import math

with open(r'd:\SalesAndStockManagement\.understand-anything\tmp\ua-file-extract-results-8.json', 'r', encoding='utf8') as f:
    data = json.load(f)

with open(r'd:\SalesAndStockManagement\.understand-anything\tmp\ua-file-analyzer-input-8.json', 'r', encoding='utf8') as f:
    input_data = json.load(f)

batchImportData = input_data.get('batchImportData', {})
neighborMap = input_data.get('neighborMap', {})

nodes = []
edges = []

file_info = {
    "backend/src/services/inventory.service.ts": {"summary": "Service quản lý hàng tồn kho, xử lý kiểm kê, tính toán giá vốn và điều chỉnh kho.", "tags": ["service", "inventory", "business-logic", "data-model"]},
    "backend/src/services/notification.service.ts": {"summary": "Service gửi thông báo và quản lý các loại cảnh báo trong hệ thống.", "tags": ["service", "notification", "utility"]},
    "backend/src/services/posting.service.ts": {"summary": "Service quản lý hạch toán và ghi nhận các bút toán kế toán.", "tags": ["service", "accounting", "business-logic"]},
    "backend/src/services/product.service.ts": {"summary": "Service quản lý thông tin sản phẩm, danh mục, và giá bán.", "tags": ["service", "product", "business-logic", "data-model"]},
    "backend/src/services/profile.service.ts": {"summary": "Service quản lý thông hồ sơ người dùng và cửa hàng.", "tags": ["service", "profile", "user-management"]},
    "backend/src/services/sales.service.ts": {"summary": "Service quản lý quy trình bán hàng, xử lý đơn hàng và thanh toán.", "tags": ["service", "sales", "business-logic", "transaction"]},
    "backend/src/services/shop-member.service.ts": {"summary": "Service quản lý thành viên trong cửa hàng và phân quyền truy cập.", "tags": ["service", "authorization", "user-management"]},
    "backend/src/services/shop-role.service.ts": {"summary": "Service định nghĩa và quản lý vai trò của người dùng trong hệ thống cửa hàng.", "tags": ["service", "authorization", "role-management"]},
    "backend/src/services/sms.service.ts": {"summary": "Service tích hợp và xử lý việc gửi tin nhắn SMS.", "tags": ["service", "sms", "integration"]},
    "backend/src/services/supplier.service.ts": {"summary": "Service quản lý thông tin nhà cung cấp và lịch sử giao dịch.", "tags": ["service", "supplier", "data-model"]},
    "backend/src/services/system.service.ts": {"summary": "Service xử lý các tác vụ hệ thống và cấu hình chung.", "tags": ["service", "system", "configuration"]},
    "backend/src/services/tax.service.ts": {"summary": "Service tính toán thuế, xử lý các nghiệp vụ liên quan đến thuế quan và hóa đơn.", "tags": ["service", "tax", "accounting"]},
    "backend/src/shop/entities.ts": {"summary": "Định nghĩa các entity liên quan đến cửa hàng, thông tin cấu hình và chi tiết shop.", "tags": ["entity", "database", "data-model"]},
    "backend/src/supplier/entities.ts": {"summary": "Định nghĩa entity cho nhà cung cấp và các thông tin liên hệ liên quan.", "tags": ["entity", "database", "data-model"]},
    "backend/src/system/audit-log.subscriber.ts": {"summary": "Subscriber cơ sở dữ liệu để tự động ghi log các thay đổi và thao tác hệ thống.", "tags": ["subscriber", "audit-log", "database", "event-handler"]},
    "backend/src/system/entities.ts": {"summary": "Định nghĩa các entity hệ thống bao gồm hồ sơ cửa hàng, nhật ký hoạt động và hóa đơn.", "tags": ["entity", "database", "data-model"]},
    "fix_imports.js": {"summary": "Script hỗ trợ tự động sửa lỗi đường dẫn import trong mã nguồn.", "tags": ["script", "utility", "build-system"]},
    "fix_imports2.js": {"summary": "Script hỗ trợ phụ để sửa lỗi và điều chỉnh định dạng import code.", "tags": ["script", "utility", "build-system"]},
    "ios/Flutter/AppFrameworkInfo.plist": {"summary": "Tệp cấu hình plist cho framework ứng dụng Flutter trên nền tảng iOS.", "tags": ["configuration", "ios", "flutter", "mobile"]},
    "ios/Flutter/Debug.xcconfig": {"summary": "Cấu hình build debug của Xcode cho dự án Flutter iOS.", "tags": ["configuration", "ios", "build-system", "debug"]},
    "ios/Flutter/Release.xcconfig": {"summary": "Cấu hình build release của Xcode cho dự án Flutter iOS.", "tags": ["configuration", "ios", "build-system", "release"]},
    "ios/Runner.xcodeproj/project.pbxproj": {"summary": "Tệp tin project chính của Xcode chứa các cấu hình build, liên kết thư viện cho iOS.", "tags": ["configuration", "ios", "xcode", "project-file"]},
    "ios/Runner.xcodeproj/project.xcworkspace/contents.xcworkspacedata": {"summary": "Định nghĩa cấu trúc workspace Xcode cho runner iOS.", "tags": ["configuration", "ios", "xcode", "workspace"]},
    "ios/Runner.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist": {"summary": "Cấu hình kiểm tra workspace Xcode của môi trường phát triển.", "tags": ["configuration", "ios", "xcode"]},
    "ios/Runner.xcodeproj/project.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings": {"summary": "Cài đặt workspace Xcode quy định các thiết lập chia sẻ chung cho iOS.", "tags": ["configuration", "ios", "xcode"]}
}

for file_result in data.get('results', []):
    path = file_result['path']
    fileCategory = file_result.get('fileCategory', 'code')
    lines = file_result.get('nonEmptyLines', 0)
    
    node_type = 'file'
    if fileCategory == 'config':
        node_type = 'config'
    elif fileCategory == 'docs':
        node_type = 'document'
    elif fileCategory == 'infra':
        node_type = 'service' if ('docker' in path.lower() or 'manifest' in path.lower()) else ('resource' if 'tf' in path.lower() else 'pipeline')
    elif fileCategory == 'data':
        node_type = 'table' if 'sql' in path.lower() else ('schema' if 'schema' in path.lower() or 'graphql' in path.lower() else 'endpoint')
        
    if path.endswith('.pbxproj') or path.endswith('.xcworkspacedata') or path.endswith('.xcsettings'):
        node_type = 'config'

    complexity = 'complex' if lines > 200 else ('moderate' if lines > 50 else 'simple')
    info = file_info.get(path, {"summary": "Tệp tin mã nguồn.", "tags": ["file"]})
    
    file_node_id = f"{node_type}:{path}"
    nodes.append({
        "id": file_node_id,
        "type": node_type,
        "name": os.path.basename(path),
        "filePath": path,
        "summary": info["summary"],
        "tags": info["tags"],
        "complexity": complexity
    })
    
    if path in batchImportData:
        for imp in batchImportData[path]:
            edges.append({"source": file_node_id, "target": f"file:{imp}", "type": "imports", "direction": "forward", "weight": 0.7})
            
    for func in file_result.get('functions', []):
        func_lines = func['endLine'] - func['startLine'] + 1
        is_exported = any(e['name'] == func['name'] for e in file_result.get('exports', []))
        if func_lines >= 10 or is_exported:
            func_id = f"function:{path}:{func['name']}"
            nodes.append({
                "id": func_id,
                "type": "function",
                "name": func['name'],
                "summary": f"Hàm {func['name']} xử lý nghiệp vụ.",
                "tags": ["function", "business-logic"],
                "complexity": "moderate" if func_lines > 20 else "simple",
                "lineRange": [func['startLine'], func['endLine']]
            })
            edges.append({"source": file_node_id, "target": func_id, "type": "contains", "direction": "forward", "weight": 1.0})
            if is_exported:
                edges.append({"source": file_node_id, "target": func_id, "type": "exports", "direction": "forward", "weight": 0.8})

    for cls in file_result.get('classes', []):
        cls_lines = cls['endLine'] - cls['startLine'] + 1
        num_methods = len(cls.get('methods', []))
        is_exported = any(e['name'] == cls['name'] for e in file_result.get('exports', []))
        if cls_lines >= 20 or num_methods >= 2 or is_exported:
            cls_id = f"class:{path}:{cls['name']}"
            nodes.append({
                "id": cls_id,
                "type": "class",
                "name": cls['name'],
                "summary": f"Lớp {cls['name']} quản lý dữ liệu và logic.",
                "tags": ["class", "data-model"],
                "complexity": "complex" if cls_lines > 100 else ("moderate" if cls_lines > 50 else "simple"),
                "lineRange": [cls['startLine'], cls['endLine']]
            })
            edges.append({"source": file_node_id, "target": cls_id, "type": "contains", "direction": "forward", "weight": 1.0})
            if is_exported:
                edges.append({"source": file_node_id, "target": cls_id, "type": "exports", "direction": "forward", "weight": 0.8})

print(f"Total nodes: {len(nodes)}")
print(f"Total edges: {len(edges)}")

os.makedirs(r'd:\SalesAndStockManagement\.understand-anything\intermediate', exist_ok=True)

parts = math.ceil(max(len(nodes) / 60, len(edges) / 120))
if parts <= 1:
    with open(r'd:\SalesAndStockManagement\.understand-anything\intermediate\batch-8.json', 'w', encoding='utf8') as f:
        json.dump({"nodes": nodes, "edges": edges}, f, indent=2, ensure_ascii=False)
    print("Written 1 parts.")
else:
    sorted_files = sorted(file_result['path'] for file_result in data.get('results', []))
    chunk_size = math.ceil(len(sorted_files) / parts)
    
    for k in range(parts):
        chunk_files = set(sorted_files[k*chunk_size : (k+1)*chunk_size])
        
        def get_file_path(node):
            if 'filePath' in node:
                return node['filePath']
            else:
                return node['id'].split(':')[1]

        part_nodes = [n for n in nodes if get_file_path(n) in chunk_files]
        part_node_ids = set(n['id'] for n in part_nodes)
        part_edges = [e for e in edges if e['source'] in part_node_ids]
        
        with open(fr'd:\SalesAndStockManagement\.understand-anything\intermediate\batch-8-part-{k+1}.json', 'w', encoding='utf8') as f:
            json.dump({"nodes": part_nodes, "edges": part_edges}, f, indent=2, ensure_ascii=False)
    print(f"Written {parts} parts.")
